scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -644,8 +646,16 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
    * [[PairSCollectionFunctions.reduceByKey]] will provide much better performance.
    * @group transform
    */
-  def groupBy[K: Coder](f: T => K): SCollection[(K, Iterable[T])] =
+  def groupBy[K: Coder](f: T => K): SCollection[(K, Iterable[T])] = {
+    if (!context.isTest && thisWasCalledExternally) {
+      logger.warn(
+        "groupBy will materialize all values for a key to a single worker," +
+          " which is a very common cause of memory issues." +
+          " Consider using `PairSCollectionFunctions.aggregateByKey` or `PairSCollectionFunctions.reduceByKey` instead."
```

Comment:
```suggestion
          " Consider using aggregateByKey/reduceByKey on a keyed SCollection instead."
```
I'd suggest this wording to match the style used by other warnings in the `com.spotify.scio.values` package.